runtime/lua/vim/lsp/_folding_range.lua

```diff
@@ -0,0 +1,400 @@
+        -- which is not the desired behavior.
+        local prev_level = row_level[start_row - 1]
+        if prev_level and prev_level[2] == '<' then
+          row_level[start_row] = { prev_level[1] - 1 }
+        end
+      elseif row < 0 then
+        vim._list_remove(row_level, start_row, start_row + math.abs(row) - 1)
+      end
+    end,
+  })
+  api.nvim_create_autocmd('LspDetach', {
+    group = augroup_setup,
+    buffer = bufnr,
+    callback = function(args)
+      ---@type integer
+      local client_id = args.data.client_id
+      bufstates[bufnr].client_ranges[client_id] = nil
+
+      ---@type vim.lsp.Client[]
+      local clients = vim
+        .iter(vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_foldingRange }))
+        ---@param client vim.lsp.Client
+        :filter(function(client)
+          return client.id ~= client_id
+        end)
+        :totable()
+      if #clients == 0 and api.nvim_buf_is_loaded(bufnr) then
+        bufstates[bufnr] = {
+          client_ranges = {},
+          row_level = {},
+          row_kinds = {},
+          row_text = {},
+        }
+      end
+
+      foldupdate(bufnr)
+    end,
+  })
+  api.nvim_create_autocmd('LspAttach', {
+    group = augroup_setup,
+    buffer = bufnr,
+    callback = function(args)
+      local client = assert(vim.lsp.get_client_by_id(args.data.client_id))
+      request(bufnr, client)
+    end,
+  })
+  api.nvim_create_autocmd('LspNotify', {
+    group = augroup_setup,
+    buffer = bufnr,
+    callback = function(args)
+      local client = assert(vim.lsp.get_client_by_id(args.data.client_id))
+      if
+        client:supports_method(ms.textDocument_foldingRange, bufnr)
+        and (
+          args.data.method == ms.textDocument_didChange
+          or args.data.method == ms.textDocument_didOpen
+        )
+      then
+        request(bufnr, client)
+      end
+    end,
+  })
+
+  request(bufnr)
+
+  return bufstates[bufnr]
+end
+
+-- The "garbage collector" for `bufstate` and its associated event hooks.
+-- See the comment above `augroup_setup` function for more details.
+--
+-- It will be fine without this, as the bufstate and event hooks will also
+-- be cleared when the buffer is unloaded.
+-- api.nvim_create_autocmd('OptionSet', {
+--   group = augroup_setup,
+--   pattern = 'foldexpr',
+--   callback = function()
+--     -- Sould find a more reliable way to detect changes in `foldexpr`.
+--     if vim.v.option_old == 'v:lua.vim.lsp.foldexpr()' then
+--       for bufnr in pairs(bufstates) do
+--         if
+--           ---@param winid integer
+--           vim.iter(vim.fn.win_findbuf(bufnr)):all(function(winid)
+--             return vim.wo[winid].foldexpr ~= 'v:lua.vim.lsp.foldexpr()'
+--           end)
+--         then
+--           bufstates[bufnr] = nil
+--           api.nvim_clear_autocmds({ buffer = bufnr, group = augroup_setup })
+--         end
+--       end
+--     end
+--   end,
+-- })
```

Comment:
This block can be removed for now. If we add it then it shouldn't try and compare the value of foldexpr.